### PR TITLE
Fix #387: Concept card linking

### LINKED ISSUE #387: Support showing concept cards in the exploration player
We need to add support for adding links to concept cards and displaying them in app with a dialog overlay that shows the concept card.

### PR DESCRIPTION
**Pull request by rt4914**
## Explanation

Fixes #387 
This PR adds the concept card linking inside exploration player. This has been done by using #422 
Thanks to @BenHenning for #422 

TAG information (as confirmed with @seanlip ):

`<oppia-noninteractive-skillreview skill_id-with-value="<[skillId]>" text-with-value="<[linkText]>"></oppia-noninteractive-skillreview>`

## How to test this:
Approach 1. Make `HtmlParserTestActivity` as your launcher activity.
Approach 2. Start `Meaning of Equal Parts` exploration. First answer is `D. last option`. In second state, answer 3/2 and you should see a `refresher lesson` text.

## Pending Issues:
- [ ] LiTagHandler is not getting used.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: utility/src/main/java/org/oppia/util/parser/HtmlParser.kt [hunk 1]
```diff
@@ -37,17 +45,23 @@ class HtmlParser private constructor(
...
 
+    // https://stackoverflow.com/a/8662457
+    if (supportsLinks) {
+      htmlContentTextView.movementMethod = LinkMovementMethod.getInstance()
+    }
+
     val imageGetter = urlImageParserFactory.create(htmlContentTextView, entityType, entityId, imageCenterAlign)
 
-    val htmlSpannable = HtmlCompat.fromHtml(htmlContent, HtmlCompat.FROM_HTML_MODE_LEGACY, imageGetter, LiTagHandler()) as Spannable
+    val htmlSpannable = CustomHtmlContentHandler.fromHtml(
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:60]: Earlier LiTagHandler was getting used to modify the bullets but now it is not getting used. Not sure how to introduce this back in the implementation. Need suggestions from @BenHenning

File: app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt [hunk 2]
```diff
@@ -2,14 +2,18 @@ package org.oppia.app.testing
...
 import org.oppia.app.activity.InjectableAppCompatActivity
+import org.oppia.app.topic.conceptcard.ConceptCardFragment
 import org.oppia.util.parser.HtmlParser
 import javax.inject.Inject
 
+private const val CONCEPT_CARD_DIALOG_FRAGMENT_TAG = "CONCEPT_CARD_FRAGMENT"
+
 /** This is a dummy activity to test Html parsing. */
-class HtmlParserTestActivity : InjectableAppCompatActivity() {
+class HtmlParserTestActivity : InjectableAppCompatActivity(), HtmlParser.CustomOppiaTagActionListener {
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt:16]: @BenHenning Can you confirm whether this testing approach is correct or not?

Comment (BenHenning) [reply app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt:16]: This seems sensible.

File: utility/src/main/java/org/oppia/util/parser/HtmlParser.kt (hunk 1 above)
Comment (BenHenning) [reply utility/src/main/java/org/oppia/util/parser/HtmlParser.kt:60]: Could we create another custom tag handler for ``li`` tags? I'm wondering if we can leverage the existing tag handler machinery to also make the list tag adjustment work correctly.

**Comment by BenHenning** [comment]
Thanks for doing this @rt4914! As general feedback, I suggest splitting this into two PRs: one for the custom tag handler machinery & dedicated tests, and a second for actually hooking it up to the concept card & tests for that. It's generally easier to reason about & verify correctness of code when you split apart the introduction of complex code from its integration points.

**Comment by rt4914** [comment]
@veena14cs @nikitamarysolomanpvt While reviewing this PR, do not focus on code because I will split this PR. But you can definelity check the implementation by the running the application.

**Comment by veena14cs** [comment]
> @veena14cs @nikitamarysolomanpvt While reviewing this PR, do not focus on code because I will split this PR. But you can definelity check the implementation by the running the application.

Okay. So far everything works properly. Assign me once it is ready for review.

**Review by nikitamarysolomanpvt — COMMENTED** [review]
I have done a review on html parser, Please assign me back after the implementation and changes suggested by ben is complete.

**Comment by BenHenning** [comment]
@rt4914 is this ready for review? Also, were tests added (per the reason #422 was never sent for review)?

**Comment by rt4914** [comment]
Closing this PR as @BenHenning would be working on this issue and also this branch is out-of-date.

**Pull request closed without merging**